gax/src/main/java/com/google/api/gax/rpc/ServerStreamingAttemptCallable.java

```diff
@@ -0,0 +1,376 @@
+ *   <li>A future representing the end state of the inner RPC is passed to the outer {@link
+ *       RetryingFuture}.
+ *   <li>All messages received from the inner {@link ServerStreamingCallable} are recorded by the
+ *       {@link StreamResumptionStrategy}.
+ *   <li>All messages received from the inner {@link ServerStreamingCallable} are forwarded to the
+ *       outer {@link ResponseObserver}.
+ *   <li>Upon RPC completion (either success or failure) are communicated to the outer {@link
+ *       RetryingFuture}.
+ *   <li>If the {@link RetryingFuture} decides to resume the RPC, it will invoke {@link #call()},
+ *       which will consult the {@link StreamResumptionStrategy} for the resuming request.
+ *   <li>Process restarts at step 5.
+ *   <li>Once the {@link RetryingFuture} decides to stop the retry loop, it will notify the outer
+ *       {@link ResponseObserver}.
+ * </ol>
+ *
+ * <p>This class is meant to be used as middleware between an outer {@link ResponseObserver} and an
+ * inner {@link ServerStreamingCallable}. As such it follows the general threading model of {@link
+ * ServerStreamingCallable}s:
+ *
+ * <ul>
+ *   <li>{@code onStart} must be called in the same thread that invoked {@code call}
+ *   <li>The outer {@link ResponseObserver} can call {@code request()} and {@code cancel()} on this
+ *       class' {@link StreamController} from any thread
+ *   <li>The inner callable will serialize calls to {@code onResponse()}, {@code onError()} and
+ *       {@code onComplete}
+ * </ul>
+ *
+ * <p>With this model in mind, this class only needs to synchronize access data that is shared
+ * between: the outer {@link ResponseObserver} (via this class' {@link StreamController}) and the
+ * inner {@link ServerStreamingCallable}: pendingRequests, cancellationCause and the current
+ * innerController.
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <RequestT> request type
+ * @param <ResponseT> response type
+ */
+final class ServerStreamingAttemptCallable<RequestT, ResponseT> implements Callable<Void> {
+  private final Object lock = new Object();
+
+  private final Watchdog<ResponseT> watchdog;
+
+  private final ServerStreamingCallable<RequestT, ResponseT> innerCallable;
+  private final StreamResumptionStrategy<RequestT, ResponseT> resumptionStrategy;
+  private final RequestT initialRequest;
+  private ApiCallContext context;
+  private final ResponseObserver<ResponseT> outerObserver;
+
+  // Start state
+  private boolean autoFlowControl = true;
+  private boolean isStarted;
+
+  // Outer state
+  @GuardedBy("lock")
+  private Throwable cancellationCause;
+
+  @GuardedBy("lock")
+  private int pendingRequests;
+
+  private RetryingFuture<Void> outerRetryingFuture;
+
+  // Internal retry state
+  private int numAttempts;
+
+  @GuardedBy("lock")
+  private StreamController innerController;
+
+  private boolean seenSuccessSinceLastError;
+  private NonCancellableFuture<Void> innerAttemptFuture;
```

Comment:
So if I understand this stuff right, the `RetryingFuture` instance (called in this class `outerRetryingFuture`) will not be exposed to the user of this client lib. In contrast to the `UnaryCallable` case when users directly receives `ApiFuture` from a `call()`, for streaming the retrying future is instantiated in `RetryingServerStreamingCallable#call()` method, which is `void` and does not return/expose the instantiated future.  Essentially for `RetryingServerStreamingCallable` the `ServerStream` object is what `ApiFuture` is for `UnaryCallable` (which makes me wonder if `ServerStream` can just implement `ApiFuture` to make the stuff look really similar, and probably have even better integration with the rest of gax, but it is a topic for another discussion).

If the above is true, and the RetryingFuture is not exposed to the user, then the user cannot control the flow by calling `cancel()` on the future (or its attempt), the actual control is done via streaming utilities (`StreamController/Observer`). Thus there is no need to use `NonCancellableFuture` here, instead a regular `SettableApiFuture` can be used (if cancellation should be forbidden, it should be done in the controller implementation instead). This allows to keep `NonCancellableFuture` methods package-private and plus  keeps the method names meaningful (public method called `<doStuff>Privately` sounds odd).

Reply:
replaced with a SettableApiFuture. Feel free to open up an issue to discuss making ResponseObservers implement a Future